### PR TITLE
feat(Text): introduce dir attribute to support RTL

### PR DESCRIPTION
**Pull request by kuzhelov**
This proposal is aimed on introducing proper text rendering in RTL mode. Its idea consists of two points:
- introduce `dir="auto"` attribute value to the text rendered by `Text` component
- from now on ensure that all components that are about rendering text information as part of their presentation should use `Text` component to wrap this textual part

### CODE REVIEW

**Comment by brianespinosa** [comment]
> As another option, could we just make dir=auto a default prop for all components? From a usability perspective, it would be ideal if we didn't require every string to use the Text component.

Yes. And lots of additional markup just to have RTL support in other components.

If we add `dir` prop to all components, it might also make sense to later make some sort of RTLProvider component that uses the context API where we can define the dir in the provider and all components will now have access to it.

**Comment by levithomason** [comment]
>If we add dir prop to all components, it might also make sense to later make some sort of RTLProvider component that uses the context API where we can define the dir in the provider and all components will now have access to it.

Indeed, we actually have an `rtl` prop right now in the Provider itself.  It determines whether the regular CSS in JS renderer is used or the RTL CSS in JS renderer.

**Comment by jurokapsiar** [comment]
Levi, do I understand correctly that after you proposed change, the ``dir='auto'`` would appear on each DOM rendered by the UIComponent? I don't think we want that. Sites typically use ``dir='rtl'`` on the ``<html>`` or ``<body>`` element to provide the default direction of the page. This will cause all elements to flip from left to right (typical example is a left-side pane becoming right-side, switching order of ok/cancel buttons and so on). On the elements that contain text we actually want to use ``dir='auto'`` which will then tell the browser to detect the direction based on the text itself.

**Comment by levithomason** [comment]
Can we define what "elements that contain text" means exactly?

### Just Text
```jsx
<span>Text only</span>
```
This element has text only, no child elements.

### Text > empty inline child
```jsx
<span>Text only with <i class="an icon"></i></span>
```
This element has text, but also has a `i` element which is common for icons.

### Text > inline child > text
```jsx
<span>Text only with <span>more text</span>.
```
This element has text, but also has a child element which contains only text.

### Text > button child > text
```jsx
<span>Text only with <button>technically text?</button>.
```
Are buttons also considered to be text?

***

Nearly every component contains text as some descendant.  Where exactly should we be adding the `dir` attribute for proper accessibility in the tree?

**Comment by levithomason** [comment]
The concern I have had with this PR is that it seems to conflate the styling purpose of the Text component with the RTL concerns.  The Text component is about styling text.  RTL to me seems to encompass more than just text and therefore perhaps its own mechanism or component.

>Text inside of the button might need dir='auto' for example in cases where you have team names as button texts. In that case the direction of the text itself need to be different depending on the content language.

Would the markup then look like this?:

```html
<button dir='auto'>
  Join Team
</button>
```

If we introduce the changes proposed in the PR, we'd end up with this:

```html
<button>
  <span dir='auto'>Join Team</span>
</button>
```

**Comment by jurokapsiar** [comment]
From RTL point of view both examples are equal. However, the second one renders one more element which is unnecessary. Is there a generic way to achieve 2? And for that case, please note that we will also need to support
```
<button><span dir='auto'>Join</span><span rtl='auto'>{team name in arabic/hebrew}</span><button>
```

**Comment by levithomason** [comment]
@jurokapsiar can you please post the final update here on this PR so we can implement it?  Thanks...

**Comment by jurokapsiar** [comment]
In general, we need to get to a state where the text nodes are inside of elements that have `dir='auto'` attribute. This allows the browser to detect the text direction.

If the text is composed of multiple parts (typically a translation with parameters), each of the parts needs to be separated in `<span dir='auto'>` element. This is however out of scope for Stardust and should be handled in the i18n layer of the application.

Requirements:
- use `Text` component to render any text inside of standard Stardust components
- `Text` component renders `<span dir='auto>{content}</span>` if the type of content is a string
- if user provides HTML/React nodes as an input for the component content, the user is responsible for adding `dir='auto'` by himself. This needs to be documented
- revisit applying of `truncateStyle` in `buttonStyles.ts` - it needs to be applied even if the content contains `<span>` or other elements

**Comment by kuzhelov** [comment]
button styles are left for the separate PR as this is the issue not directly related to this PR. Also, we need the following PR to be merged first to ensure that styles will be applied to children `Text`: #496 

Dedicated issue for `Button` styles is created: https://github.com/stardust-ui/react/issues/520 .

**Review by jurokapsiar — APPROVED** [review]

**Review by sophieH29 — APPROVED** [review]

File: src/components/Text/Text.tsx [hunk 1]
```diff
@@ -85,9 +85,14 @@ class Text extends UIComponent<Extendable<TextProps>, any> {
 
   renderComponent({ ElementType, classes, rest }): React.ReactNode {
     const { children, content } = this.props
+
+    const hasChidlren = childrenExist(children)
+
+    const maybeDirAuto = !hasChidlren && typeof content === 'string' ? { dir: 'auto' } : null
```
Comment (miroslavstastny) [inline src/components/Text/Text.tsx]: `hasChidlren` -> `hasChildren`

**Review by miroslavstastny — APPROVED** [review]

**Pull request merged**